src/provider/apiresponse.h

```diff
@@ -0,0 +1,54 @@
+#ifndef ERD_PROXY_API_RESPONSE_H
+#define ERD_PROXY_API_RESPONSE_H
+
+#include "json/json.hpp"
+#include "../utils/errors.h"
+
+class ErdGenericApiResponse
+{
+public:
+    explicit ErdGenericApiResponse(std::string rawData)
+    {
+        try
+        {
+            m_response = nlohmann::json::parse(rawData);
+        }
+        catch (...)
+        {
+            throw std::invalid_argument(ERROR_MSG_JSON_SERIALIZED + rawData);
+        }
+    }
+
+    std::string getError() const
+    {
+        std::string ret;
+
+        if (m_response.contains("error"))
+            if(m_response["error"] != "")
+                ret = m_response["error"];
+
+       return ret;
+    }
+
+    std::string getCode() const
+    {
+        if (!m_response.contains("code"))
+            throw std::invalid_argument(ERROR_MSG_JSON_KEY_NOT_FOUND + "code");
+
+        return m_response["code"];
+    }
+
+    template <class T>
+    T getData() const
+    {
+        if (!m_response.contains("data"))
+            throw std::invalid_argument(ERROR_MSG_JSON_KEY_NOT_FOUND + "data");
+
+        return m_response["data"];
```

Comment:
So here, the indexing operator `[...]` is generic as well, right? 

Not sure, is it possible to rewrite `getData()` to include the logic in `getDataIfValid()` as well? And perhaps also rename it to `extractData()`? `getPayload()`?

Or perhaps move the logic from `getDataIfValid` to the constructor (`parseResponse() / parseJson()`)?

How about this piece of code (just an idea)?

```
auto response = ErdGenericApiResponse.parseJson("...");
response.raiseOnError();
auto foo = response.getPayload();
```

Reply:
> So here, the indexing operator [...] is generic as well, right?

In `ErdGenericApiResponse` -> YES, it is generic and returns a template class.
However, in `ProxyProvider` we always expect data (if valid) to be a json.

> How about this piece of code (just an idea)?

I've applied your suggestion.
However, I've decided to include those 3 lines of code in a function to avoid duplicate code.
So now, we have this internal function, with your proposed functionality:
```
nlohmann::json getPayLoad(wrapper::http::Result const &res)
    {
        ErdGenericApiResponse response = internal::parse(res);
        response.checkSuccessfulOperation();

        return response.getData<nlohmann::json>();
    }
```